ops/testing.py

```diff
@@ -564,10 +615,26 @@ def add_relation(self, relation_name: str, remote_app: str) -> int:
         self._backend._relation_ids_map.setdefault(relation_name, []).append(rel_id)
         self._backend._relation_names[rel_id] = relation_name
         self._backend._relation_list_map[rel_id] = []
+
+        # leaders can write their app databag but not read it
+        def _can_read_local_app_data():
+            if self._backend._hook_is_running:
+                return not self._backend.is_leader()
+            return True
```

Comment:
ach, so bad... 
I went a few times back and forth with that logic because I got jumbled several times with keeping in mind who could read what. Good catch!